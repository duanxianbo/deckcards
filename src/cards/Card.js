import React from 'react';
import PropTypes from 'prop-types';
import injectSheet from 'react-jss';
import {colorTypeMap} from '../config';

const cardIconSylte = {
  color: ({type}) => colorTypeMap[type],
  content: ({type}) => `'${type}'`,
  position: 'absolute'
};

const styles = {
  cardContainer: {
    marginRight: 10,
    position: 'relative',
    width: 50,
    height: 70,
    borderRadius: '10px',
    background: '#fff',
    boxShadow: '3px 3px 7px rgba(0,0,0,0.3)',
    '&:before': {...cardIconSylte, top: 2, left: 2},
    '&:after': {...cardIconSylte, bottom: 2, right: 2},
    display: 'flex',
    alignItems: 'center'
  },
  cardContent: {
    flex: 1,
    textAlign: 'center',
    font: 'Georgia, Times New Roman, serif'
  }
};

function Card({element, classes}) {
  return (
    <div className={classes.cardContainer}>
      <div className={classes.cardContent}>{element}</div>
    </div>
  );
}

Card.propTypes = {
  element: PropTypes.string,
  classes: PropTypes.object
};


export default injectSheet(styles)(Card);
