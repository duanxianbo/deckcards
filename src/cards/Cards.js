import React from 'react';
import PropTypes from 'prop-types';
import Card from './Card';
import Typography from '@material-ui/core/Typography';

function Cards({cards}) {
  return (
    cards && cards.length ? cards.map(({id, ...others}) => (
      <Card key={id} {...others}/>
    )) :
      (<Typography variant="subtitle2" color="inherit" noWrap>
        No more cards here
      </Typography>)
  );
}

Cards.propTypes = {
  cards: PropTypes.array
};


export default (Cards);
