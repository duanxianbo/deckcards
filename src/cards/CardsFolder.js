import React from 'react';
import PropTypes from 'prop-types';
import Cards from './Cards';
import injectSheet from 'react-jss';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';

const styles = {
  cardsList: {
    justifyContent: ({cards}) => cards.length === 0 && 'center',
    alignItems: ({cards}) => cards.length === 0 && 'center'
  },
  actionButton: {
    height: 36,
    width: 250
  }
};

function CardsFolder({cards, cardAction, actionLabel, classes}) {
  return (
    <Grid
      item
      container
      spacing={32}
      direction={'column'}>
      <Grid item container justify={'center'}>
        <Button className={classes.actionButton} onClick={cardAction}>{actionLabel}</Button>
      </Grid>
      <Grid item container className={classes.cardsList}>
        <Cards cards={cards}/>
      </Grid>
    </Grid>
  );
}

CardsFolder.propTypes = {
  cards: PropTypes.array,
  cardAction: PropTypes.func,
  actionLabel: PropTypes.string,
  classes: PropTypes.object
};


export default injectSheet(styles)(CardsFolder);
