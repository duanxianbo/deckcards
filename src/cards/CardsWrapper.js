import React from 'react';
import PropTypes from 'prop-types';
import CardsFolder from './CardsFolder';
import injectSheet from 'react-jss';

const styles = {
  container: {
    display: 'flex',
    flexWrap: 'wrap'
  },
  cardsFolder: {
    display: 'flex',
    flexBasis: '50%',
    flexShrink: 0,
    flexWrap: 'wrap',
    flex: 1,
    minWidth: '300px'
  }
};

function CardsWrapper({cardsDeck, cardsDealed, dealOneCardAction, shuffleAction, classes}) {
  return (
    <div className={classes.container}>
      <div className={classes.cardsFolder}>
        <CardsFolder
          cards={cardsDeck}
          cardAction={shuffleAction}
          actionLabel={'shuffle'}
          id="cardsDeck"/>
      </div>
      <div className={classes.cardsFolder}>
        <CardsFolder
          cards={cardsDealed}
          cardAction={dealOneCardAction}
          actionLabel={'deal one card'}
          id="cardsDealed"/>
      </div>
    </div>
  );
}


CardsWrapper.propTypes = {
  cardsDeck: PropTypes.array,
  cardsDealed: PropTypes.array,
  dealOneCardAction: PropTypes.func,
  shuffleAction: PropTypes.func,
  classes: PropTypes.object
};

export default injectSheet(styles)(CardsWrapper);
