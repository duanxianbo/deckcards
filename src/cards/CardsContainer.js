import React from 'react';
import {connect} from 'react-redux';
import {shuffle, dealOneCard} from '../actions';
import CardsWrapper from './CardsWrapper';

function CardsContainer(props) {
  return <CardsWrapper {...props}/>;
}

export default connect(
  state => state,
  (dispatch) => ({
    shuffleAction: () => dispatch(shuffle()),
    dealOneCardAction: () => dispatch(dealOneCard())
  }),
)(CardsContainer);
