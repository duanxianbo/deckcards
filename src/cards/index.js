import Card from './Card';
import CardsFolder from './CardsFolder';
import Cards from './Cards';
import CardsWrapper from './CardsWrapper';
import CardsContainer from './CardsContainer';

export {Card, Cards, CardsFolder, CardsContainer, CardsWrapper};
