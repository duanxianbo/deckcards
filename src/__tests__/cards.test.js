import React from 'react';
import {render, mount} from 'enzyme';
import {Card, Cards, CardsFolder, _CardsContainer} from '../cards';

const cards = [{
  id: 1,
  element: 'A',
  type: '♥'
}, {
  id: 2,
  element: '1',
  type: '♥'
}, {
  id: 3,
  element: '4',
  type: '♣'
}];

describe('Cards basic', () => {

  it('should render correctly one card', () => {
    const component = render(<Card element={'A'} type={'♣'} />);

    expect(component).toMatchSnapshot();
  });


  it('should render cards', () => {
    const component = render(<Cards cards={cards} />);

    expect(component).toMatchSnapshot();
  });
});


describe('Cards actions', () => {
  it('should perform card actions', () => {
    const cardAction = jest.fn();
    const component = mount(<CardsFolder cards={cards} cardAction={cardAction} actionLabel={'test'}/>);

    component.find('button').simulate('click');
    expect(cardAction).toHaveBeenCalled();

    component.unmount();
  });

  it('should perform card actions', () => {
    const dealOneCardAction = jest.fn();
    const shuffle = jest.fn();

    const component = mount(
      <_CardsContainer
        cardsDeck={cards.slice(0, 1)}
        cardsDealed={cards.slice(1)}
        dealOneCardAction={dealOneCardAction}
        shuffleAction={shuffle}
      />);

    component.find('#cardsDeck button').simulate('click');
    expect(shuffle).toHaveBeenCalled();

    component.find('#cardsDealed button').simulate('click');
    expect(dealOneCardAction).toHaveBeenCalled();

    component.unmount();
  });
});
