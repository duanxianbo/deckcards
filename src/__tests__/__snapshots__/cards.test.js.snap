// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards basic should render cards 1`] = `
Array [
  <div
    class="Card-cardContainer-0-1-1 Card-cardContainer-0-1-4"
  >
    <div
      class="Card-cardContent-0-1-2"
    >
      A
    </div>
  </div>,
  <div
    class="Card-cardContainer-0-1-1 Card-cardContainer-0-1-5"
  >
    <div
      class="Card-cardContent-0-1-2"
    >
      1
    </div>
  </div>,
  <div
    class="Card-cardContainer-0-1-1 Card-cardContainer-0-1-6"
  >
    <div
      class="Card-cardContent-0-1-2"
    >
      4
    </div>
  </div>,
]
`;

exports[`Cards basic should render correctly one card 1`] = `
<div
  class="Card-cardContainer-0-1-1 Card-cardContainer-0-1-3"
>
  <div
    class="Card-cardContent-0-1-2"
  >
    A
  </div>
</div>
`;
